backend/llm/BrainPickingOpenAIFunctions/BrainPickingOpenAIFunctions.py

```diff
@@ -0,0 +1,177 @@
+                {"role": "assistant", "content": chat.assistant},
+            ]
+        ]
+
+    def _get_context(self, question: str) -> str:
+        """
+        Retrieve documents related to the question
+        """
+        logger.info("Getting context")
+        vector_store = CustomSupabaseVectorStore(
+            self.supabase_client,
+            self.embeddings,
+            table_name="vectors",
+            user_id=self.user_email,
+        )
+
+        return vector_store.similarity_search(query=question)
+
+    def _construct_prompt(
+        self, question: str, useContext: bool = False, useHistory: bool = False
+    ) -> List[Dict[str, str]]:
+        """
+        Constructs a prompt given a question, and optionally include context and history
+        """
+        logger.info("Constructing prompt")
+        system_messages = [
+            {
+                "role": "system",
+                "content": "Your name is Quivr. You are a second brain. A person will ask you a question and you will provide a helpful answer. Write the answer in the same language as the question.If you don't know the answer, just say that you don't know. Don't try to make up an answer.our main goal is to answer questions about user uploaded documents. Unless basic questions or greetings, you should always refer to user uploaded documents by fetching them with the get_context function.",
+            }
+        ]
+
+        if useHistory:
+            logger.info("Adding chat history to prompt")
+            history = self._get_chat_history()
+            system_messages.append(
+                {"role": "system", "content": "Previous messages are already in chat."}
+            )
+            system_messages.extend(history)
+
+        if useContext:
+            logger.info("Adding chat context to prompt")
+            chat_context = self._get_context(question)
+            context_message = f"Here is chat context: {chat_context if chat_context else 'No document found'}"
+            system_messages.append({"role": "user", "content": context_message})
+
+        system_messages.append({"role": "user", "content": question})
+
+        return system_messages
+
+    def get_answer(self, question: str) -> str:
+        """
+        Main function to get an answer for the given question
+        """
+        logger.info("Getting answer")
+        functions = [
+            {
+                "name": "get_history",
+                "description": "Used to get the chat history between the user and the assistant",
+                "parameters": {"type": "object", "properties": {}},
+            },
+            {
+                "name": "get_context",
+                "description": "Used for retrieving documents related to the question to help answer the question",
+                "parameters": {"type": "object", "properties": {}},
+            },
+        ]
+
+        # First, try to get an answer using just the question
```

Comment:
That means the user calls twice the API most of the time for one request on quivr, non ?